jfix-zookeeper/src/main/kotlin/ru/fix/zookeeper/instance/registry/ServiceInstanceIdRegistry.kt

```diff
@@ -0,0 +1,196 @@
+class ServiceInstanceIdRegistry(
+        private val curatorFramework: CuratorFramework,
+        private val instanceIdGenerator: InstanceIdGenerator,
+        private val config: ServiceInstanceIdRegistryConfig,
+        profiler: Profiler
+) : AutoCloseable {
+    lateinit var lockIdentity: LockIdentity
+        private set
+    lateinit var instanceIdPath: String
+        private set
+    lateinit var instanceId: String
+        private set
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(ServiceInstanceIdRegistry::class.java)
+    }
+
+    private val lockManager: PersistentExpiringLockManager = PersistentExpiringLockManager(
+            curatorFramework,
+            DynamicProperty.of(
+                    PersistentExpiringLockManagerConfig(
+                            reservationPeriod = config.disconnectTimeout.dividedBy(2),
+                            expirationPeriod = config.disconnectTimeout.dividedBy(3),
+                            lockProlongationInterval = config.disconnectTimeout.dividedBy(4),
+                            acquiringTimeout = Duration.ofSeconds(1)
+                    )
+            ),
+            profiler
+    )
+
+    init {
+        curatorFramework
+                .connectionStateListenable
+                .addListener(
+                        ConnectionStateListener { _, newState ->
+                            when (newState) {
+                                ConnectionState.RECONNECTED -> reconnect()
+                                else -> Unit
+                            }
+                        })
+        initServiceRegistrationPath()
+        initInstanceId()
+    }
+
+    private fun initServiceRegistrationPath() {
+        try {
+            curatorFramework.create()
+                    .creatingParentsIfNeeded()
+                    .forPath(config.serviceRegistrationPath, byteArrayOf())
+        } catch (e: KeeperException.NodeExistsException) {
+            logger.debug("Node ${config.serviceRegistrationPath} is already initialized", e)
+        } catch (e: Exception) {
+            logger.error("Illegal state when create path: ${config.serviceRegistrationPath}", e)
+        }
+    }
+
+    /**
+     * Try {@link #config.countRegistrationAttempts} times to register instance.
+     * If this is client with host and port, that already contains in instance's znodes, and timeout of this lock expired,
+     * this situation seen as reconnect of instance, and previous instance id will be got.
+     * Generate instance id, this id should be in range 1..maxInstancesCount, assertion error will be thrown otherwise.
+     * If unsuccessful, then log error
+     */
+    private fun initInstanceId() {
+        for (i in 1..config.countRegistrationAttempts) {
+            val alreadyRegisteredInstanceIds = curatorFramework.children.forPath(config.serviceRegistrationPath)
+            val preparedInstanceId = resolvePreviousConnection(alreadyRegisteredInstanceIds)
```

Comment:
Why we are using host and port here? ServiceInstanceIdRegistry is a registry for services. It is possible that we will have 10 services created within single application that coexist within single process. We will have tow processes and 20 services in total.

Reply:
port and host was removed